### PR TITLE
CORTX-32261: Expose CSM APIs for Activity Tracker

### PR DESCRIPTION
**Pull request by rohitkolapkar**
Signed-off-by: Rohit Kolapkar <rohit.j.kolapkar@seagate.com>

# Pull Request
## Problem Statement
-   [Expose CSM REST APIs for Activity Tracker ](https://jts.seagate.com/browse/CORTX-32261)

## Design
-   [API Specs](https://seagate-systems.atlassian.net/wiki/spaces/PUB/pages/1063649497/Activity+Tracker+HLD#Activity-Tracker%3A-Management-API-Specifications)
-   \[x] Coding conventions are followed and code is consistent

-   \[ \] Unit and System Tests are added
-   \[ \] Test Cases cover Happy Path, Non-Happy Path and Scalability
-   \[ \] Testing was performed with RPM

-   \[ \] Interface change (if any) are documented
-   \[ \] Side effects on other features (deployment/upgrade)
-   \[ \] Dependencies on other component(s)

-   \[x] JIRA number/GitHub Issue added to PR
-   \[x] PR is self reviewed
-   \[x] Jira and state/status is updated and JIRA is updated with PR link
-   \[x] Check if the description is clear and explained

-   \[ \] Changes done to WIKI / Confluence page / Quick Start Guide

##Dev Testing

###Happy Path
1. Create an activity
image

2. Fetch an activity using activity_id
image

3. Update the activity
image

###Un-Happy Path
1.  API validations
image

image

image

2. If activity does not exist - update and fetch should fail
image

# Swagger Updated
image

### CODE REVIEW

**Comment by rohitkolapkar** [comment]
@pawankumarsrivastava  @kupranay This PR is dependent on Utils PR - https://github.com/Seagate/cortx-utils/pull/834

File: csm/core/controllers/activities.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+        activity_id = self.request.match_info[const.ID]
+        path_params = {const.ID: activity_id}
+        try:
+            schema = UpdateActivitySchema()
+            request_params = schema.load(await self.request.json())
+            request_body = {**path_params, **request_params}
+            Log.debug(f"Handling update activity PATCH request"
+                      f" request body: {request_body}")
+        except json.decoder.JSONDecodeError:
+            raise InvalidRequest("Could not parse request body, invalid JSON received.")
```
Comment (abhijit1patil) [inline csm/core/controllers/activities.py:106]: you can use const.JSON_ERROR

**Review by abhijit1patil — CHANGES_REQUESTED** [review]

File: csm/core/controllers/activities.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        """POST REST implementation for creating a new activity."""
+        Log.info(f"Handling create an activity POST request"
+                 f" user_id: {self.request.session.credentials.user_id}")
+        try:
+            schema = CreateActivitySchema()
+            request_body = schema.load(await self.request.json())
+            Log.debug(f"Handling create an activity POST request"
+                      f" request body: {request_body}")
+        except json.decoder.JSONDecodeError:
+            raise InvalidRequest("Could not parse request body, invalid JSON received.")
```
[the remark this answers is not in the document]
Comment (rohitkolapkar) [reply csm/core/controllers/activities.py]: yes, thanks for reminding.

[the remark this answers is not in the document]
Comment (rohitkolapkar) [reply csm/core/controllers/activities.py]: updated.

File: csm/core/controllers/activities.py (hunk 1 above)
Comment (rohitkolapkar) [reply csm/core/controllers/activities.py]: updated.

**Comment by rohitkolapkar** [comment]
1N Deployment is failing because of dependency over Utils PR.
Custom Ci - passed
https://eos-jenkins.colo.seagate.com/job/GitHub-custom-ci-builds/job/generic/job/custom-ci/6688/console

File: csm/core/controllers/activities.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    """Update activity schema validation class."""
+
+    progress = fields.Int(validate=validate.Range(min=0, max=100), data_key=const.PROGRESS,
+        required=True)
+    status = fields.Str(data_key=const.STATUS_LITERAL,
+        validate=validate.OneOf(Status.list()), required=True)
+    status_description = fields.Str(data_key=const.STATUS_DESC, required=True)
+
+
+@CsmView._app_routes.view("/api/v2/activities")
```
Comment (pawankumarsrivastava) [inline csm/core/controllers/activities.py:48]: /api/v2/activities or /api/v2/activity ?

**Review by pawankumarsrivastava — CHANGES_REQUESTED** [review]

File: csm/core/controllers/activities.py (hunk 3 above)
Comment (rohitkolapkar) [reply csm/core/controllers/activities.py:48]: as per REST API norms - /api/v2/activities

**Review by pawankumarsrivastava — APPROVED** [review]

**Review by abhijit1patil — CHANGES_REQUESTED** [review]

File: csm/core/controllers/activities.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    description = fields.Str(data_key=const.DESCRIPTION, required=True)
+
+
+class UpdateActivityBaseSchema(ValidateSchema):
+    """Update activity base schema validation class."""
+
+    status = fields.Str(data_key=const.STATUS_LITERAL, required=True,
+                           validate=validate.OneOf(const.SUPPORTED_ACTIVITY_STATUS))
+    status_description = fields.Str(data_key=const.STATUS_DESC, required=True)
+    pct_progress = fields.Int(data_key=const.PCT_PROGRESS)
```
[the remark this answers is not in the document]
Comment (rohitkolapkar) [reply csm/core/controllers/activities.py:43]: This is the base schema, there are more schemas based  on status value.
FYI, required=true covers allow_none=False and missing=None is not required here. 
Also please go through API specs for update api.

File: csm/core/services/activities.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Affero General Public License for more details.
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <https://www.gnu.org/licenses/>.
+# For any questions about this software or licensing,
+# please email opensource@seagate.com or cortx-questions@seagate.com.
+
+# CORTX-CSM: CORTX Management web and CLI interface.
```
[the remark this answers is not in the document]
Comment (rohitkolapkar) [reply csm/core/services/activities.py:16]: removed

**Review by abhijit1patil — APPROVED** [review]

**Review by pawankumarsrivastava — CHANGES_REQUESTED** [review]
Please update the following - 
1. Swagger Doc
2. CSM Endpoints Doc - https://seagate-systems.atlassian.net/wiki/spaces/PUB/pages/994836509/CORTX+Management+Endpoints

**Review by pawankumarsrivastava — APPROVED** [review]

**Pull request merged**